### DISCUSSION TITLE
records limitation on findMany

### DISCUSSION THREAD

**Discussion by cybernadinou** [prisma/prisma#9115] [Q&A] [answered]
Hello everyone !!

My table `item` got for now around 50 000 records.

Here is my request  : 
```ts
const listOffers = await prisma.offer.findMany({
  include: {
    item: true,
  },
})
```
When I execute the above request, I got the following error : 

```
PrismaClientKnownRequestError3 [PrismaClientKnownRequestError]: Can't reach database server at `localhost`:`54328`

Please make sure your database server is running at `localhost`:`54328`.
    at cb (/my-path/my-appy/node_modules/@prisma/client/runtime/index.js:36494:17)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  code: 'P1001',
  clientVersion: '2.30.3',
  meta: { database_host: 'localhost', database_port: 54328 }
}
```

But when I put a limitation (more than `34749` will throw the error above) : 
```ts
const listOffers = await prisma.offer.findMany({
  include: {
    model: { include: { family: { include: { department: true } } } },
    item: true,
  },
  take: 34749,
})
```
The request is working fine. 

I also tried some stuff like below :

```ts
const listOffers = await prisma.offer.findMany({
  include: {
    model: { include: { family: { include: { department: true } } } },
    item: true,
  },
  skip: 20000,
  take: 20000,
})
```
To be sure it was not my records or any information in my database that was crashing, and everything works fine.

Does it means  prisma can't return more than 34749 records or do I miss a configuration ?  
I'm asking here because I didn't find anything in the prisma documentation nor on postrgresql documentation about any limitation.

Thanks for your help !

**Comment by janpio** [comment]
Can you activate debugging and logging please and post the full output you are getting? https://pris.ly/logging and https://pris.ly/debugging

**Reply by cybernadinou** [reply]
Hello !
Thx for your answer !!

Here is my file with logs
[all_logs.txt](https://github.com/prisma/prisma/files/7109096/all_logs.txt)

**Reply by janpio** [reply]
Ok. Not much unexpected in there - you send the query, and you get _that_ answer from the server instead of the data.

Can you try this preview feature in your project please and report if it changes something about the error? https://www.prisma.io/docs/concepts/components/prisma-engines/query-engine/#enable-the-node-api-n-api-preview

Can you maybe dump the data of the relevant tables into a .sql file so we can reproduce this manually?

**Reply by cybernadinou** [reply]
hello !!

Thanks again for your help ! 

I put the dump of the database on this private repo and I gave you acces @janpio 
https://github.com/cybernadinou/upload-file 

I also tried the solution you told  me to by adding "nApi" to the previewFetaures : 

```
generator client {
  provider = "prisma-client-js"
  previewFeatures = ["referentialActions", "orderByRelation", "nApi"]
}
```

Unfortunately, I got the same error : 
```
PrismaClientKnownRequestError3 [PrismaClientKnownRequestError]: Can't reach database server at `localhost`:`54328`

Please make sure your database server is running at `localhost`:`54328`.
    at cb (/my-path/my-appy/node_modules/@prisma/client/runtime/index.js:36494:17)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  code: 'P1001',
  clientVersion: '2.30.3',
  meta: { database_host: 'localhost', database_port: 54328 }
}
```



I also tried the solution below of @ryands17  but same didn't work either :/

I just notice something in the log by typing something like : `cat all_logs.txt | grep error -A 1 -B 1`, we got an error on `prisma:engine`:
```
2021-09-04T02:29:30.480Z prisma:engine Active provider: postgresql
2021-09-04T02:29:31.695Z prisma:engine stdout Unknown error                    <==== THIS LINE HERE
prisma:query SELECT "public"."Offer"."externalId", "public"."Offer"."validFrom", "public"."Offer"."validThrough", "public"."Offer"."range", "public"."Offer"."third", "public"."Offer"."createdAt", "public"."Offer"."updatedAt", "public"."Offer"."modelId", "public"."Offer"."itemId" FROM "public"."Offer" WHERE 1=1 OFFSET $1
2021-09-04T02:29:32.389Z prisma:engine stdout Unknown error                    <===== HERE AGAIN 
prisma:query SELECT "public"."Model"."id", "public"."Model"."name", "public"."Model"."pixlId", "public"."Model"."superModelId", "public"."Model"."createdAt", "public"."Model"."updatedAt", "public"."Model"."familyId" FROM "public"."Model"
```

Just in case it might help :)

EDIT: I just added a partial `schema.prisma` just in case it helps you. Only tables related to this issue.
Thanks again !

**Reply by janpio** [reply]
> I just notice something in the log by typing something like : `cat all_logs.txt | grep error -A 1 -B 1`, we got an error on `prisma:engine`:

We can not reliable reproduce that one yet, but are tracking it at: https://github.com/prisma/prisma/issues/8052

**Reply by janpio** [reply]
Thanks! Can you also give @ryands17 and @pantharshit00 access please? That would be very helpful as I am super busy with something else right now. Thanks.

**Reply by cybernadinou** [reply]
Done :)
EDIT : 

Version Info:
MacOS : Catalina v10.15.7
Nodejs : v14.17.1
Npm: 7.18.1
Prisma : 2.30.2

Don't know if it matters but I'm using typescript.

**Reply by ryands17** [reply]
@cybernadinou Checked the repo you gave and it looks like Prisma errors out fetching around 35000 records:

![image](https://user-images.githubusercontent.com/19697099/132165770-d4e4e892-4f9b-4e00-aa64-dffe363d577c.png)

Here's the script I ran with the Postgres dump you provided:

```ts
yarn prisma db seed --preview-feature
```

So we would need to track this in the issue itself.

**Reply by janpio** [reply]
Bug report issue with all the info, and a _public_ reproduction script would be neat @cybernadinou 

We will probably want to look at the database server log files to see if it indeed crashes or goes away :/ @ryands17

**Reply by blevine** [reply]
It looks like Prisma is generating a `select *  from some_table where some_id in (lots_of_params) lots_of_values`  rather than just a `select * from some_table`.  Why is that?

**Reply by ryands17** [reply]
The reason is that it fetches all the id's that are connected to the relation and then it fetches those values using `in`.

**Reply by blevine** [reply]
I can't see the related schema, but it seems like this should just generate a join or a where clause using a foreign key.

**Reply by bruno02100** [reply]
I'm having this same exact problem, any work around to it?

**Comment by ryands17** [comment]
Most likely it's because Prisma batches queries in a specific limit. Adding a higher value to this environment variable should make it work:

```
QUERY_BATCH_SIZE=50000 node script.js
```

**Reply by janpio** [reply]
The server just going away and this weird error message would still be unexpected and really bad.

**Reply by bruno02100** [reply]
Is this env variable actually supported? I'm having the same problem and tried this env and didn't work.

**Comment by w8ze-devel** [comment]
I can confirm that the environment variable didn't have the expected effect.

**Comment by w8ze-devel** [comment]
Hello,

Do others people experimenting this problem ?

I remember i encounter it while loading a schema WITH relation. I just load 36318 record in a table (WITHOUT loading relations) and everything is working fine.

Plus I remember also that on the project where I encounter the problem, the environment variable didn't have the expected effect.

Thanks

**Comment by typ3castter** [comment]
I am also having this problem. Do we not have a resolution after all this time?

**Reply by typ3castter** [reply, marked as answer]
Resolved after I upgraded from Prisma 3 to 4 ( 4.12 )

**Comment by erentomioke22** [comment]
![80_generated](https://github.com/prisma/prisma/assets/169043260/07e58c89-e02d-424d-9547-cf77992ccaae)
fggggggggggggggggggg jggggggggggg

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
